### ISSUE TITLE
[BUG] PreCache does not persist accross sessions

### ISSUE THREAD

**Issue by themadmrj** [jhomlala/betterplayer#506]
**History check**
Since I initially wrote the preCache PR, I know it doesn't :)

**Describe the bug**
The pre-caching system works fine in the same sessions. However, if you quit the app, that is quit the last screen or force kill, and then return, the same video will be downloaded again.

**To Reproduce**
See Describe the bug to reproduce the simple steps.

**Example code*
No code required, same behaviour for any video (on android ofc, since no pre-caching on iOS atm).

**Expected behavior**
The cached video should persist across sessions.

<details><summary>**Flutter doctor**</summary>
[√] Flutter (Channel stable, 2.0.3, on Microsoft Windows [Version 10.0.19041.985], locale fr-FR)
    • Flutter version 2.0.3 at E:\Flutter
    • Framework revision 4d7946a68d (2 months ago), 2021-03-18 17:24:33 -0700
    • Engine revision 3459eb2436
    • Dart version 2.12.2

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\Users\MrJ\AppData\Local\Android\sdk
    • Platform android-30, build-tools 29.0.2
    • Java binary at: E:\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 4.1.0)
    • Android Studio at E:\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] Connected device (2 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 90.0.4430.212
    • Edge (web)   • edge   • web-javascript • Microsoft Edge 90.0.818.56
</details>

**Better Player version**
- Version: 0.0.68

**Smartphone (please complete the following information):**
All android devices

**Additional context**
I would try to find the cause myself, but atm I really lack the time. My guess is, `SimpleCache` is somehow not shared or persisted as it should and the following session creates a new cache instead of reading from the old one.

**Comment by jhomlala** [comment]
@themadmrj I can confirm that I can reproduce this issue on my side. I've checked if ExoPlayer keeps files between app sessions and I can confirm that files are not being deleted. Seems like ExoPlayer doesn't load videos from that dir after re-init. 
I've did some research but I couldn't find any solution.

**Comment by themadmrj** [comment]
According to [this link](https://github.com/google/ExoPlayer/issues/8797), it should work, so there has to be a parameter somewhere which we missed

**Comment by themadmrj** [comment]
I did some testing, the solution is fairly easy: a key is never used. I use temp urls (maybe you do too), so my urls change all the time. When I set a custom key, both when caching and when reading, it works exactly as expected. I managed to read the video offline in another session and also checked that the cache does not get rewritten for the same file. If you could allow passing a custom cache key from dart, that would be nice :)

**Comment by themadmrj** [comment]
Also, regarding iOS pre-caching, [this library](https://github.com/huangguiyang/LLVideoPlayer/tree/master/LLVideoPlayer/CacheSupport) is also based on AV Player and has pre loading feature. I'm not sure if it's still maintained, so a fairly good idea would be to get the pre loading code from there and integrate it here!

**Comment by jhomlala** [comment]
@themadmrj Thanks for that check. Can you provide some sample code for that or link to documentation?

**Comment by themadmrj** [comment]
> @themadmrj Thanks for that check. Can you provide some sample code for that or link to documentation?

For Android, first `CacheWorker`, something like `DataSpec dataSpec = new DataSpec(uri, 0, preCacheSize).buildUpon().setKey(myKey).build();` or use the Builder. And then in `BetterPlayer`, you have to change the `buildMediaSource` method. I did this to test:
```
            case C.TYPE_OTHER:
                return new ProgressiveMediaSource.Factory(mediaDataSourceFactory,
                        new DefaultExtractorsFactory())
                        .setCustomCacheKey(key)
                        .setDrmSessionManager(drmSessionManager)
                        .createMediaSource(MediaItem.fromUri(uri));
```

But `.setCustomCacheKey(key)` seems deprecated, so there's probably a better way to do it.

**Comment by themadmrj** [comment]
@jhomlala the docs apparrently are outdated. The solution is in the inline docs:
```
    /**
     * Sets the optional custom cache key (only used for progressive streams).
     *
     * <p>If {@link #setUri} is passed a non-null {@code uri}, the custom cache key is used to
     * create a {@link PlaybackProperties} object. Otherwise it will be ignored.
     */
    public Builder setCustomCacheKey(@Nullable String customCacheKey) {
      this.customCacheKey = customCacheKey;
      return this;
    }
```

I found this in `MediaItem.java`

**Comment by jhomlala** [comment]
Fixed in 0.0.69.
